File: frontend/src/services/voice/functionExecutors.ts
import config from '../../config';
import type { EventbriteEvent, SessionInfo, SearchResult } from './types';

/**
 * Result of a function execution
 */
export interface FunctionResult {
  success: boolean;
  output: string;
}

/**
 * Execute the meeting_notes function
 */
export async function executeMeetingNotes(args: {
  action: string;
  filter?: string;
  query?: string;
  top_k?: number;
  session_filter?: string;
}): Promise<FunctionResult> {
  const action = args.action;

  if (action === 'list_sessions') {
    return await listSessions(args.filter);
  } else if (action === 'search') {
    return await searchMeetingNotes(args.query || '', args.top_k || 5, args.session_filter);
  } else {
    return {
      success: false,
      output: `Unknown action: ${action}. Use 'list_sessions' or 'search'.`
    };
  }
}

/**
 * List available sessions
 */
async function listSessions(filter?: string): Promise<FunctionResult> {
  try {
    const params = new URLSearchParams();
    if (filter) {
      params.append('filter', filter);
    }

    const url = `${config.apiBaseUrl}/v1/sessions${params.toString() ? '?' + params : ''}`;
    const response = await fetch(url);

    if (!response.ok) {
      return {
        success: false,
        output: `Failed to list sessions: ${response.statusText}`
      };
    }

    const data = await response.json();

    if (data.sessions && data.sessions.length > 0) {
      const output = `Found ${data.sessions.length} session(s):\n` +
        data.sessions.map((s: SessionInfo, i: number) =>
          `${i + 1}. ${s.session_info} (${s.chunk_count} chunks)`
        ).join('\n');
      return { success: true, output };
    } else {
      return { success: true, output: 'No sessions found matching the filter.' };
    }
  } catch (error) {
    return {
      success: false,
      output: 'Error listing sessions: ' + (error instanceof Error ? error.message : 'Unknown error')
    };
  }
}

/**
 * Search meeting notes
 */
async function searchMeetingNotes(
  query: string,
  topK: number,
  sessionFilter?: string
): Promise<FunctionResult> {
  try {
    const params = new URLSearchParams({
      question: query,
      top_k: String(topK)
    });

    if (sessionFilter) {
      params.append('session_filter', sessionFilter);
    }

    const url = `${config.apiBaseUrl}/v1/search?${params}`;
    const response = await fetch(url);

    if (!response.ok) {
      return {
        success: false,
        output: `Search failed: ${response.statusText}`
      };
    }

    const results = await response.json();

    if (results.results && results.results.length > 0) {
      const output = results.results.map((r: SearchResult, i: number) =>
        `${i + 1}. [Session: ${r.session_info}, Timestamp: ${r.timestamp}, Score: ${r.score.toFixed(3)}]\n   ${r.text}`
      ).join('\n\n');
      return { success: true, output };
    } else {
      return { success: true, output: 'No relevant meeting notes found for this query.' };
    }
  } catch (error) {
    return {
      success: false,
      output: 'Error searching: ' + (error instanceof Error ? error.message : 'Unknown error')
    };
  }
}

/**
 * Extended event type with full details
 */
interface EventbriteEventFull extends EventbriteEvent {
  id: string;
  full_description?: string;
  agenda?: Array<{ time: string; title: string }>;
}

/**
 * Execute the eventbrite function (list or details)
 */
export async function executeEventbrite(args: {
  action: string;
  limit?: number;
  event_id?: string;
}): Promise<FunctionResult> {
  try {
    const action = args.action || 'list';

    if (action === 'details') {
      return await fetchEventDetails(args.event_id);
    } else {
      return await fetchEventList(args.limit || 3);
    }
  } catch (error) {
    return {
      success: false,
      output: 'Error with eventbrite: ' + (error instanceof Error ? error.message : 'Unknown error')
    };
  }
}

/**
 * Fetch list of upcoming events
 */
async function fetchEventList(limit: number): Promise<FunctionResult> {
  const url = `${config.apiBaseUrl}/v1/events?action=list&limit=${limit}`;
  const response = await fetch(url);

  if (!response.ok) {
    return {
      success: false,
      output: `Failed to fetch events: ${response.statusText}`
    };
  }

  const data = await response.json();

  if (data.events && data.events.length > 0) {
    const output = data.events.map((event: EventbriteEventFull, i: number) => {
      const parts = [`${i + 1}. **${event.name}** (ID: ${event.id})`];

      if (event.start_date) {
        let timeStr = event.start_time;
        if (event.end_time) {
          timeStr += ` - ${event.end_time}`;
        }
        parts.push(`   Date: ${event.start_date} at ${timeStr}`);
      }

      if (event.location) {
        parts.push(`   Location: ${event.location}`);
      }

      if (event.description) {
        const desc = event.description.length > 200
          ? event.description.substring(0, 200) + '...'
          : event.description;
        parts.push(`   Description: ${desc}`);
      }

      if (event.price) {
        parts.push(`   Price: ${event.price}`);
        if (event.tickets_available !== null) {
          parts.push(`   Tickets Available: ${event.tickets_available}`);
        }
      } else if (event.is_free) {
        parts.push('   Price: Free');
      }

      if (event.url) {
        parts.push(`   Register: ${event.url}`);
      }

      return parts.join('\n');
    }).join('\n\n');

    return { success: true, output };
  } else {
    return {
      success: true,
      output: 'No upcoming events found. Check back later or visit the Birmingham AI Eventbrite page.'
    };
  }
}

/**
 * Fetch full details for a specific event
 */
async function fetchEventDetails(eventId?: string): Promise<FunctionResult> {
  if (!eventId) {
    return {
      success: false,
      output: 'Error: event_id is required for details action'
    };
  }

  const url = `${config.apiBaseUrl}/v1/events?action=details&event_id=${eventId}`;
  const response = await fetch(url);

  if (!response.ok) {
    return {
      success: false,
      output: `Failed to fetch event details: ${response.statusText}`
    };
  }

  const data = await response.json();
  const event = data.event as EventbriteEventFull;

  if (!event) {
    return {
      success: false,
      output: `Event ${eventId} not found.`
    };
  }

  const parts = [`**${event.name}**`];

  if (event.start_date) {
    let timeStr = event.start_time;
    if (event.end_time) {
      timeStr += ` - ${event.end_time}`;
    }
    parts.push(`Date: ${event.start_date} at ${timeStr}`);
  }

  if (event.location) {
    parts.push(`Location: ${event.location}`);
  }

  if (event.full_description) {
    parts.push(`\n**Description:**\n${event.full_description}`);
  } else if (event.description) {
    parts.push(`\n**Description:**\n${event.description}`);
  }

  if (event.agenda && event.agenda.length > 0) {
    parts.push('\n**Agenda:**');
    for (const item of event.agenda) {
      parts.push(`  - ${item.time}: ${item.title}`);
    }
  }

  if (event.price) {
    parts.push(`\nPrice: ${event.price}`);
    if (event.tickets_available !== null) {
      parts.push(`Tickets Available: ${event.tickets_available}`);
    }
  } else if (event.is_free) {
    parts.push('\nPrice: Free');
  }

  if (event.url) {
    parts.push(`\nRegister: ${event.url}`);
  }

  return { success: true, output: parts.join('\n') };
}

/**
 * Execute a function by name
 */
export async function executeFunction(
  name: string,
  argsJson: string
): Promise<FunctionResult> {
  try {
    const args = JSON.parse(argsJson || '{}');

    switch (name) {
      case 'meeting_notes':
        return await executeMeetingNotes(args);
      case 'get_upcoming_events':
        return await executeEventbrite({ action: 'list', limit: args.limit });
      case 'eventbrite':
        return await executeEventbrite(args);
      default:
        return {
          success: false,
          output: `Unknown function: ${name}`
        };
    }
  } catch (error) {
    return {
      success: false,
      output: 'Error executing function: ' + (error instanceof Error ? error.message : 'Unknown error')
    };
  }
}

File: pi-voice-client/src/function_executors.py
"""
Function executors for OpenAI Realtime API function calling.

Handles executing functions called by the AI and returning results.
"""

import logging
import json
from typing import Dict, Any
import httpx
from .config import config

logger = logging.getLogger(__name__)


async def execute_meeting_notes(args: Dict[str, Any]) -> Dict[str, Any]:
    """
    Execute meeting_notes function.
    
    Args:
        args: Function arguments with 'action', 'filter', 'query', 'top_k', 'session_filter'
    
    Returns:
        Dictionary with 'success' (bool) and 'output' (str) keys
    """
    action = args.get("action", "")
    
    if action == "list_sessions":
        return await _list_sessions(args.get("filter"))
    elif action == "search":
        return await _search_meeting_notes(
            args.get("query", ""),
            args.get("top_k", 5),
            args.get("session_filter")
        )
    else:
        return {
            "success": False,
            "output": f"Unknown action: {action}. Use 'list_sessions' or 'search'."
        }


async def _list_sessions(filter_text: str = None) -> Dict[str, Any]:
    """List available meeting sessions."""
    try:
        params = {}
        if filter_text:
            params["filter"] = filter_text
        
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                f"{config.API_BASE_URL}/v1/sessions",
                params=params
            )
            
            if response.status_code != 200:
                return {
                    "success": False,
                    "output": f"Failed to list sessions: {response.status_code}"
                }
            
            data = response.json()
            sessions = data.get("sessions", [])
            
            if sessions:
                output_lines = [f"Found {len(sessions)} session(s):"]
                for i, s in enumerate(sessions, 1):
                    session_info = s.get("session_info", "Unknown")
                    chunk_count = s.get("chunk_count", 0)
                    output_lines.append(f"{i}. {session_info} ({chunk_count} chunks)")
                return {"success": True, "output": "\n".join(output_lines)}
            else:
                return {
                    "success": True,
                    "output": "No sessions found matching the filter."
                }
    
    except Exception as e:
        logger.error(f"Error listing sessions: {e}")
        return {
            "success": False,
            "output": f"Error listing sessions: {str(e)}"
        }


async def _search_meeting_notes(
    query: str, top_k: int = 5, session_filter: str = None
) -> Dict[str, Any]:
    """Search meeting notes."""
    try:
        params = {
            "question": query,
            "top_k": str(top_k)
        }
        if session_filter:
            params["session_filter"] = session_filter
        
        async with httpx.AsyncClient(timeout=15.0) as client:
            response = await client.get(
                f"{config.API_BASE_URL}/v1/search",
                params=params
            )
            
            if response.status_code != 200:
                return {
                    "success": False,
                    "output": f"Search failed: {response.status_code}"
                }
            
            data = response.json()
            results = data.get("results", [])
            
            if results:
                output_lines = []
                for i, r in enumerate(results, 1):
                    session = r.get("session_info", "Unknown")
                    timestamp = r.get("timestamp", "")
                    score = r.get("score", 0.0)
                    text = r.get("text", "")
                    output_lines.append(
                        f"{i}. Session: {session}, Time: {timestamp}, "
                        f"Score: {score:.3f}\n   {text}"
                    )
                return {
                    "success": True,
                    "output": "\n\n".join(output_lines)
                }
            else:
                return {
                    "success": True,
                    "output": "No relevant meeting notes found for this query."
                }
    
    except Exception as e:
        logger.error(f"Error searching meeting notes: {e}")
        return {
            "success": False,
            "output": f"Error searching: {str(e)}"
        }


async def execute_eventbrite(args: Dict[str, Any]) -> Dict[str, Any]:
    """
    Execute eventbrite function.
    
    Args:
        args: Function arguments with 'action', 'limit', 'event_id'
    
    Returns:
        Dictionary with 'success' (bool) and 'output' (str) keys
    """
    action = args.get("action", "list")
    
    if action == "details":
        return await _get_event_details(args.get("event_id"))
    else:
        return await _list_events(args.get("limit", 3))


async def _list_events(limit: int = 3) -> Dict[str, Any]:
    """List upcoming events."""
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                f"{config.API_BASE_URL}/v1/events",
                params={"action": "list", "limit": limit}
            )
            
            if response.status_code != 200:
                return {
                    "success": False,
                    "output": f"Failed to fetch events: {response.status_code}"
                }
            
            data = response.json()
            events = data.get("events", [])
            
            if events:
                output_lines = []
                for i, event in enumerate(events, 1):
                    name = event.get("name", "Unknown Event")
                    event_id = event.get("id", "")
                    date = event.get("start_date", "")
                    time = event.get("start_time", "")
                    location = event.get("location", "")
                    
                    lines = [f"{i}. {name} (ID: {event_id})"]
                    if date:
                        end_time = event.get("end_time", "")
                        time_str = f"{time} - {end_time}" if end_time else time
                        lines.append(f"   Date: {date} at {time_str}")
                    if location:
                        lines.append(f"   Location: {location}")
                    
                    description = event.get("description", "")
                    if description:
                        desc = description[:150] + "..." if len(description) > 150 else description
                        lines.append(f"   Description: {desc}")
                    
                    # Add price info
                    if event.get("price"):
                        lines.append(f"   Price: {event.get('price')}")
                        if event.get("tickets_available") is not None:
                            lines.append(f"   Tickets Available: {event.get('tickets_available')}")
                    elif event.get("is_free"):
                        lines.append("   Price: Free")
                    
                    # Add URL
                    if event.get("url"):
                        lines.append(f"   Register: {event.get('url')}")
                    
                    output_lines.append("\n".join(lines))
                
                return {
                    "success": True,
                    "output": "\n\n".join(output_lines)
                }
            else:
                return {
                    "success": True,
                    "output": "No upcoming events found."
                }
    
    except Exception as e:
        logger.error(f"Error listing events: {e}")
        return {
            "success": False,
            "output": f"Error fetching events: {str(e)}"
        }


async def _get_event_details(event_id: str = None) -> Dict[str, Any]:
    """Get details for a specific event."""
    if not event_id:
        return {
            "success": False,
            "output": "Error: event_id is required for details action"
        }
    
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                f"{config.API_BASE_URL}/v1/events",
                params={"action": "details", "event_id": event_id}
            )
            
            if response.status_code != 200:
                return {
                    "success": False,
                    "output": f"Failed to fetch event details: {response.status_code}"
                }
            
            data = response.json()
            event = data.get("event")
            
            if not event:
                return {
                    "success": False,
                    "output": f"Event {event_id} not found."
                }
            
            lines = [event.get("name", "Unknown Event")]
            
            date = event.get("start_date")
            time = event.get("start_time")
            end_time = event.get("end_time")
            if date:
                time_str = f"{time} - {end_time}" if end_time else time
                lines.append(f"Date: {date} at {time_str}")
            
            location = event.get("location")
            if location:
                lines.append(f"Location: {location}")
            
            # Use full description if available
            description = event.get("full_description") or event.get("description")
            if description:
                lines.append(f"\nDescription:\n{description}")
            
            # Add agenda if available
            agenda = event.get("agenda")
            if agenda and isinstance(agenda, list) and len(agenda) > 0:
                lines.append("\nAgenda:")
                for item in agenda:
                    if isinstance(item, dict):
                        item_time = item.get("time", "")
                        item_title = item.get("title", "")
                        lines.append(f"  - {item_time}: {item_title}")
            
            # Add price info
            if event.get("price"):
                lines.append(f"\nPrice: {event.get('price')}")
                if event.get("tickets_available") is not None:
                    lines.append(f"Tickets Available: {event.get('tickets_available')}")
            elif event.get("is_free"):
                lines.append("\nPrice: Free")
            
            # Add URL
            if event.get("url"):
                lines.append(f"\nRegister: {event.get('url')}")
            
            return {"success": True, "output": "\n".join(lines)}
    
    except Exception as e:
        logger.error(f"Error getting event details: {e}")
        return {
            "success": False,
            "output": f"Error fetching event details: {str(e)}"
        }


async def execute_function(name: str, args_json: str) -> Dict[str, Any]:
    """
    Execute a function by name.
    
    Args:
        name: Function name
        args_json: JSON string of function arguments
    
    Returns:
        Dictionary with 'success' (bool) and 'output' (str) keys
    """
    try:
        args = json.loads(args_json or "{}")
        
        if name == "meeting_notes":
            return await execute_meeting_notes(args)
        elif name in ["get_upcoming_events", "eventbrite"]:
            return await execute_eventbrite(args)
        else:
            return {
                "success": False,
                "output": f"Unknown function: {name}"
            }
    
    except json.JSONDecodeError as e:
        logger.error(f"Failed to parse function arguments: {e}")
        return {
            "success": False,
            "output": f"Error parsing arguments: {str(e)}"
        }
    except Exception as e:
        logger.error(f"Error executing function {name}: {e}")
        return {
            "success": False,
            "output": f"Error executing function: {str(e)}"
        }
